### PR TITLE
Add modules of SSN pipeline.

### PR DESCRIPTION
**Pull request by tony2016uestc**

### CODE REVIEW

File: mmaction/datasets/pipelines/loading.py [hunk 1]
```diff
@@ -270,6 +271,245 @@ def _get_test_clips(self, num_frames):
...
+        Returns:
+            np.ndarray: Sampled frame indices in train mode.
+        """
+        avg_interval = (valid_length + 1) // num_segments
+        if avg_interval > 0:
+            base_offsets = np.arange(num_segments) * avg_interval
+            offsets = base_offsets + np.random.randint(
+                avg_interval, size=num_segments)
+        else:
+            offsets = np.zeros((num_segments, ))
```
Comment (innerlee) [inline mmaction/datasets/pipelines/loading.py:346]: dtype?

File: mmaction/datasets/pipelines/loading.py [hunk 2]
```diff
@@ -270,6 +271,245 @@ def _get_test_clips(self, num_frames):
...
+
+        Returns:
+            np.ndarray: Sampled frame indices in validation mode.
+        """
+        if valid_length >= num_segments:
+            avg_interval = valid_length / float(num_segments)
+            base_offsets = np.arange(num_segments) * avg_interval
+            offsets = (base_offsets + avg_interval / 2.0).astype(np.int32)
+        else:
+            offsets = np.zeros((num_segments, ))
```
Comment (innerlee) [inline mmaction/datasets/pipelines/loading.py:370]: dtype?

File: mmaction/datasets/pipelines/loading.py [hunk 3]
```diff
@@ -270,6 +271,245 @@ def _get_test_clips(self, num_frames):
...
+        stages: starting, course and ending stage.
+
+        Args:
+            proposal (object): The proposal object.
+            num_frames (int): Total number of frame in the video.
+
+        Returns:
+            np.ndarray: Sampled frame indices in train mode.
+        """
+        start_frame = proposal.start_frame + 1
```
Comment (innerlee) [inline mmaction/datasets/pipelines/loading.py:387]: why +1?

Comment (tony2016uestc) [reply mmaction/datasets/pipelines/loading.py:387]: Because the interval of the proposal is (proposal.start_frame, proposal.end_frame]

Comment (innerlee) [reply mmaction/datasets/pipelines/loading.py:387]: why did they choose this convention? can we change it to the python-style interval?

**Comment by innerlee** [comment]
kindly ping

**Comment by dreamerlin** [comment]
Kindly ping.

**Comment by innerlee** [comment]
The interval convention part has not changed.

File: mmaction/datasets/pipelines/loading.py [hunk 4]
```diff
@@ -270,6 +271,253 @@ def _get_test_clips(self, num_frames):
...
+
+        Args:
+            num_frames (int): Total number of frame in the video.
+
+        Returns:
+            np.ndarray: Sampled frame indices in test mode.
+        """
+        ori_clip_len = self.clip_len * self.frame_interval
+        return np.arange(
+            0, num_frames - ori_clip_len, self.test_interval, dtype=int) + 1
```
Comment (dreamerlin) [inline mmaction/datasets/pipelines/loading.py:467]: This `+1` seems a start index which may have conflict with the #Line 508

File: mmaction/datasets/pipelines/loading.py [hunk 5]
```diff
@@ -88,7 +89,7 @@ def _get_train_clips(self, num_frames):
             ratio = (num_frames - ori_clip_len + 1.0) / self.num_clips
             clip_offsets = np.around(np.arange(self.num_clips) * ratio)
         else:
-            clip_offsets = np.zeros((self.num_clips, ))
+            clip_offsets = np.zeros((self.num_clips, ), dtype=np.int32)
```
Comment (innerlee) [inline mmaction/datasets/pipelines/loading.py:92]: any reason to choose int32?

Comment (tony2016uestc) [reply mmaction/datasets/pipelines/loading.py:92]: Will it be better to change it to int8？

Comment (innerlee) [reply mmaction/datasets/pipelines/loading.py:92]: .....

np.int is ok

Comment (tony2016uestc) [reply mmaction/datasets/pipelines/loading.py:92]: OK.

File: mmaction/datasets/pipelines/loading.py [hunk 6]
```diff
@@ -270,6 +271,253 @@ def _get_test_clips(self, num_frames):
...
+                valid interval.
+            num_segments (int): Total number of segments.
+
+        Returns:
+            np.ndarray: Sampled frame indices in validation mode.
+        """
+        if valid_length >= num_segments:
+            avg_interval = valid_length / float(num_segments)
+            base_offsets = np.arange(num_segments) * avg_interval
+            offsets = (base_offsets + avg_interval / 2.0).astype(np.int)
```
Comment (innerlee) [inline mmaction/datasets/pipelines/loading.py:374]: will `offsets = base_offsets + avg_interval // 2` be ints?

Comment (innerlee) [reply mmaction/datasets/pipelines/loading.py:374]: emm no

File: mmaction/datasets/pipelines/loading.py [hunk 7]
```diff
@@ -270,6 +271,253 @@ def _get_test_clips(self, num_frames):
...
+
+        Args:
+            proposal (object): The proposal object.
+            num_frames (int): Total number of frame in the video.
+
+        Returns:
+            np.ndarray: Sampled frame indices in train mode.
+        """
+        start_frame = proposal.start_frame
+        end_frame = proposal.end_frame - 1
```
Comment (innerlee) [inline mmaction/datasets/pipelines/loading.py:394]: `end_frame = proposal.end_frame`

File: mmaction/datasets/pipelines/loading.py [hunk 8]
```diff
@@ -270,6 +271,253 @@ def _get_test_clips(self, num_frames):
...
+        ori_clip_len = self.clip_len * self.frame_interval
+
+        duration = end_frame - start_frame + 1
```
Comment (innerlee) [inline mmaction/datasets/pipelines/loading.py:397]: `duration = end_frame - start_frame`

File: mmaction/datasets/pipelines/loading.py [hunk 9]
```diff
@@ -270,6 +271,253 @@ def _get_test_clips(self, num_frames):
...
+        assert duration != 0
+        valid_length = duration - ori_clip_len
+
+        valid_starting = max(1,
```
Comment (innerlee) [inline mmaction/datasets/pipelines/loading.py:401]: is the calculation assumes that
* 1 based counting
* inclusive for both starting and ending?

i was trying to move the convention of 0 based, [start, end)

**Pull request merged**